### DISCUSSION TITLE
[Question][Hw3] Semantic of VC generation of assume()

### DISCUSSION THREAD

**Discussion by jidnjs** [prosyslab-classroom/cs424-program-reasoning#430] [Q&A] [answered]
### Your Name

Jiwon Kang (강지원)

### Question

`... @assume(%conv) ...`의 검증 조건은  `... /\ %conv  /\ ...`가 되어야 할 것 같은데,

과제 3번의 [2. Verification Condition Generation] 항목에서는
`... @assume(%conv) ...`를 검증 조건으로 변환하였을 때 `... /\ (%conv > 0) /\ ...` 와 같이 된다고 예시로 주어져 있습니다.

하지만 `assume` 함수의 이름과 

> The assume function imposes a constraint which will hold afterward. 

라는 설명을 미루어 보았을 때 `... /\ %conv  /\ ...` 가 되어야 한다고 생각합니다. 

혹시나 제가 잘못 이해한 게 있나 싶어 질문 드립니다.

---

The verification condition for `... @assume(%conv) ...` seems like it should be `... /\ %conv /\ ...`.

However, in Assignment 3, under item [2. Verification Condition Generation], the example given for converting `... @assume(%conv) ...` into a verification condition is `... /\ (%conv > 0) /\ ...`.

Considering the name of the `assume` function and the description

> The assume function imposes a constraint which will hold afterward.

I believe it should be `... /\ %conv /\ ...`. 

I'm asking in case I might have misunderstood something.

**Comment by sujin0529** [comment]
안녕하세요.
질문 주신 내용은 [이전 질문](https://github.com/prosyslab-classroom/cs424-program-reasoning/discussions/324) 의 1번 답변이 도움이 될 것 같습니다. `assume` 함수의 인자 타입을 확인해보시기 바랍니다.

감사합니다.

===========================

Hello.
I think the answer 1 of the [previous question](https://github.com/prosyslab-classroom/cs424-program-reasoning/discussions/324) will be helpful for your question. Please check the parameter type of the `assume` function.

Thanks.

**Comment by jidnjs** [comment]
답변 감사합니다!
사실 과제를 하면서 `assume` 함수의 인자 타입이 `i32`라는 것을 알게 되었고, 
`... @assume(%conv) ...` 의 검증 조건을 만들 때 단순히 `%conv`를 논리값처럼 사용하는 것은 불가능하다는 것을 깨달았습니다.

추가적으로 `... /\ (%conv != 0) /\ ...` 대신 `... /\ (%conv > 0) /\ ...`를 사용한 것은 
1) `Z3` 함수들이 사용된 구현에 있어 간단함 때문인지 (전자는 `Z3.Boolean.mk_not`과 `Z3.Boolean.mk_eq` 사용, 후자는 단순히 `Z3.Arithmetic.mk_gt` 사용)
2) 아니라면 양수의 값만이 참으로 여겨지기 때문인지 궁금합니다.
--- 
Thank you for response! In fact, while working on the assignment, I realized that the argument type of the `assume` function is `i32`, and I understood that it's impossible to use `%conv` simply as a boolean value when creating the verification condition for `... @assume(%conv) ...`.

Additionally, I'm curious why `... /\ (%conv > 0) /\ ...` was used instead of `... /\ (%conv != 0) /\ ...`. Is it because:
1) It's simpler due to the implementation using `Z3` functions (the former uses `Z3.Boolean.mk_not` and `Z3.Boolean.mk_eq`, while the latter simply uses `Z3.Arithmetic.mk_gt`), or 
2) Is it because only positive values are considered true?

**Comment by sujin0529** [comment, marked as answer]
안녕하세요.
구현이 편하기 때문이라고 말씀드릴 수 있을 것 같습니다.
궁금하시다면 `eq` 와 `not` 을 사용해서 직접 확인해보시는 것도 좋을 것 같습니다 :)
참고로 `%cmp` 는 1비트의 정수 타입 값이어서 `%conv` 는 0 또는 1만 될 수 있습니다. 

감사합니다.

===========================

Hello.
Because I think using `gt` is easier to implement.
If you are curious, it might be good to check using `eq` and `not` as well :)
For your reference, `%cmp` is a 1-bit integer type value, so `%conv` can only be 0 or 1.

Thanks.
